### PR TITLE
[optimization] use replace top instead of push pop in priority #5312

### PR DESCRIPTION
**Pull request by stdpain**
## Proposed changes

referen to #5312 

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [] Bugfix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [x] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If these changes need document changes, I have updated the document
- [x] Any dependent changes have been merged

### CODE REVIEW

File: be/src/util/sort_heap.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+#include "common/compiler_util.h"
+
+namespace doris {
+
+template <typename T, typename _Sequence, typename _Compare>
+class SortingHeap {
+public:
+    SortingHeap(const _Compare& comp) : _comp(comp) {}
+
+    bool isValid() const { return !_queue.empty(); }
```
Comment (yangzhg) [inline be/src/util/sort_heap.h:33]:
```suggestion
    bool is_valid() const { return !_queue.empty(); }
```

File: be/src/exec/topn_node.h [hunk 2]
```diff
@@ -101,12 +102,8 @@ class TopNNode : public ExecNode {
...
     int64_t _num_rows_skipped;
 
-    // The priority queue will never have more elements in it than the LIMIT.  The stl
-    // priority queue doesn't support a max size, so to get that functionality, the order
-    // of the queue is the opposite of what the ORDER BY clause specifies, such that the top
-    // of the queue is the last sorted element.
-    boost::scoped_ptr<std::priority_queue<Tuple*, std::vector<Tuple*>, TupleRowComparator>>
-            _priority_queue;
+    // The priority queue will never have more elements in it than the LIMIT.      
+    boost::scoped_ptr<SortingHeap<Tuple*, std::vector<Tuple*>, TupleRowComparator>> _priority_queue;
```
Comment (yangzhg) [inline be/src/exec/topn_node.h:106]:
```suggestion
    std::unique_ptr<SortingHeap<Tuple*, std::vector<Tuple*>, TupleRowComparator>> _priority_queue;
```

**Review by HappenLee — CHANGES_REQUESTED** [review]
some code need change~

File: be/src/exec/topn_node.h [hunk 3]
```diff
@@ -101,12 +102,8 @@ class TopNNode : public ExecNode {
...
     int64_t _num_rows_skipped;
 
-    // The priority queue will never have more elements in it than the LIMIT.  The stl
-    // priority queue doesn't support a max size, so to get that functionality, the order
-    // of the queue is the opposite of what the ORDER BY clause specifies, such that the top
-    // of the queue is the last sorted element.
-    boost::scoped_ptr<std::priority_queue<Tuple*, std::vector<Tuple*>, TupleRowComparator>>
-            _priority_queue;
+    // The priority queue will never have more elements in it than the LIMIT.      
+    std::unique_ptr<SortingHeap<Tuple*, std::vector<Tuple*>, TupleRowComparator>> _priority_queue;
```
[the remark this answers is not in the document]
Comment (stdpain) [reply be/src/exec/topn_node.h:106]: SortingHeap is essentially a priority queue, but with a replace_top interface, so naming it as a priority queue is more friendly to people who come into contact with this code for the first time

File: be/src/util/sort_heap.h [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        auto child_it = begin + child_idx;
+
+        /// Check if we are in order.
+        //if (*child_it < *begin) return;
+        if (_comp(*child_it, *begin)) return;
+
+        _next_idx = 0;
+
+        auto curr_it = begin;
+        auto top(std::move(*begin));
```
Comment (HappenLee) [inline be/src/util/sort_heap.h:97]: remove unless std::move.

File: be/src/util/sort_heap.h [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+#include <algorithm>
+#include <cassert>
+#include <queue>
+#include <utility>
+
+#include "common/compiler_util.h"
+
+namespace doris {
+
+template <typename T, typename _Sequence, typename _Compare>
```
Comment (HappenLee) [inline be/src/util/sort_heap.h:28]: no need add "_" in type name

**Review by HappenLee — CHANGES_REQUESTED** [review]
some little thing need to change

**Review by HappenLee — APPROVED** [review]
@stdpain LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**